glTF-Toolkit/src/GLTFMeshSerializationHelpers.cpp

```diff
@@ -0,0 +1,1018 @@
+        }
+        else
+        {
+            // Use the uniqueness count to determine number of vertices for each primitive.
+            std::unordered_set<uint32_t> UniqueVertices;
+
+            for (size_t i = 0; i < mesh.primitives.size(); ++i)
+            {
+                const auto& p = mesh.primitives[i];
+                auto& primInfo = m_Primitives[i];
+
+                uint32_t IndexStart = (uint32_t)m_Indices.size();
+                ReadAccessor(reader, doc, p.indicesAccessorId, m_Indices, primInfo[Indices]);
+
+                // Generate the unique vertex set.
+                UniqueVertices.clear();
+                UniqueVertices.insert(m_Indices.begin() + IndexStart, m_Indices.end());
+
+                primInfo.Offset = IndexStart;
+                primInfo.IndexCount = m_Indices.size() - IndexStart;
+                primInfo.VertexCount = UniqueVertices.size(); 
+                primInfo.CopyMeta(primInfo0);
+            }
+        }
+    }
+    else
+    {
+        primInfo0.Offset = 0;
+        primInfo0.IndexCount = 0;
+        primInfo0.VertexCount = m_Positions.size();
+    }
+}
+
+void MeshInfo::Reset(void)
+{
+    m_Name.clear();
+    m_Primitives.clear();
+
+    m_Indices.clear();
+    m_Positions.clear();
+    m_Normals.clear();
+    m_Tangents.clear();
+    m_UV0.clear();
+    m_UV1.clear();
+    m_Color0.clear();
+    m_Joints0.clear();
+    m_Weights0.clear();
+
+    m_Attributes ={ 0 };
+    m_PrimFormat = PrimitiveFormat::Combine;
+}
+
+void MeshInfo::Optimize(void)
+{
+    if (!m_Attributes.Has(Indices))
+    {
+        std::cout << "Mesh '" << m_Name << "': optimize operation failed - this operation requires mesh to use indices.";
+        return;
+    }
+
+    // DirectXMesh intermediate data
+    std::vector<uint32_t> facePrims; // Mapping from face index to primitive index.
+    std::vector<uint32_t> pointReps;
+    std::vector<uint32_t> adjacency;
+    std::vector<uint32_t> dupVerts;
+    std::vector<uint32_t> faceRemap;
+    std::vector<uint32_t> vertRemap;
+
+    const size_t indexCount = m_Indices.size();
+    const size_t vertexCount = m_Positions.size();
+    const size_t faceCount = GetFaceCount();
+
+    for (size_t i = 0; i < m_Primitives.size(); ++i)
+    {
+        // Populate attribute id array with primitive index denoting this is a distinct sub-mesh.
+        size_t oldFaceCount = facePrims.size();
+        facePrims.resize(oldFaceCount + m_Primitives[i].FaceCount());
+        std::fill(facePrims.begin() + oldFaceCount, facePrims.end(), (uint32_t)i);
+    }
+
+    // Ensure intermediate buffer sizes.
+    pointReps.resize(vertexCount);
+    adjacency.resize(GetFaceCount() * 3);
+    faceRemap.resize(facePrims.size());
+    vertRemap.resize(vertexCount);
+
+    // Perform DirectXMesh optimizations
+    DirectX::GenerateAdjacencyAndPointReps(m_Indices.data(), faceCount, m_Positions.data(), vertexCount, EPSILON, pointReps.data(), adjacency.data());
```

Comment:
These all return ``HRESULT`` values, so you need to check for failure.

Reply:
It assume it's fine to do a ``DX::ThrowIfFailed()`` kind of thing.

Reply:
Cool, got that in. Am I supposed to do anything with the dupVerts output from Directx::Clean(...)?

Reply:
Yes. See what I'm doing with it in the [meshconvert](https://github.com/Microsoft/DirectXMesh/blob/master/Meshconvert/Mesh.cpp) code.